### PR TITLE
Feature/add extension manager

### LINKED ISSUE #72: Add support for extension manager
In order to support the [extension mananger](https://github.com/exasol/extension-manager/) we need to add an extension that manages the Kafka connector.

## Steps for installation
See [user guide](https://github.com/exasol/kafka-connector-extension/blob/main/doc/user_guide/user_guide.md):
* Upload adapter jar to bucketfs ✅ 
* Create three UDF scripts ✅ 
* Create Exasol table with columns `KAFKA_PARTITION DECIMAL(18, 0), KAFKA_OFFSET DECIMAL(36, 0)` ❌ 
* Run import with `IMPORT INTO <schema_name>.<table_name> FROM SCRIPT KAFKA_CONSUMER WITH ...` ❌ 
* If required: create connection with `IDENTIFIED BY 'SSL_KEY_PASSWORD=<PASSWORD>;SSL_KEYSTORE_PASSWORD=<SSLPASSWORD>;SSL_KEYSTORE_LOCATION=/buckets/bfsdefault/<BUCKET>/keystore.jks;SSL_TRUSTSTORE_PASSWORD=<TRUSTSTOREPASS>;SSL_TRUSTSTORE_LOCATION=/buckets/bfsdefault/<BUCKET>/truststore.jks'` ❌

### PR DESCRIPTION
**Pull request by pj-spoelders**
Closes #72

### CODE REVIEW

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        previousVersion.upgrade();
+        assertInstalledVersion("Kafka Connector Extension", PROJECT_VERSION);
+        verifyImportWorks();
+    }
+
+    private void assertInstalledVersion(final String expectedName, final String expectedVersion) {
+        final List<InstallationsResponseInstallation> installations = setup.client().getInstallations();
+        final InstallationsResponseInstallation expectedInstallation = new InstallationsResponseInstallation()
+                .name(expectedName).version(expectedVersion);
+        // The extension is installed twice (previous and current version), so each one returns the same installation.
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java:227]:
```suggestion
        // The extension definition is installed twice (previous and current version), so each one returns the same installation.
```

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+            final Table targetTable = schema.createTableBuilder("TARGET")
+
+                    .column("SENSOR_ID", "INTEGER").column("STATUS", "VARCHAR(10)") //
+                    .column("KAFKA_PARTITION", "DECIMAL(18, 0)").column("KAFKA_OFFSET", "DECIMAL(36, 0)").build();
+            // CREATE CONNECTION (optional, see
+            // https://github.com/exasol/kafka-connector-extension/blob/main/doc/user_guide/user_guide.md#importing-records)
+
+            executeKafkaImport(targetTable, kafkaSetup);
+
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java:255]:
```suggestion
        try {
            final Table targetTable = schema.createTableBuilder("TARGET")
                    .column("SENSOR_ID", "INTEGER").column("STATUS", "VARCHAR(10)") //
                    .column("KAFKA_PARTITION", "DECIMAL(18, 0)").column("KAFKA_OFFSET", "DECIMAL(36, 0)").build();
            executeKafkaImport(targetTable, kafkaSetup);
```

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+                " BOOTSTRAP_SERVERS = '" + modifiedBootstrapServers + "'\n" + //
+                " RECORD_KEY_FORMAT = 'string'" + "\n" + //
+                " RECORD_VALUE_FORMAT = 'string'" + "\n" + //
+                // " SCHEMA_REGISTRY_URL = '" + kafkaConnection + "'\n" + //is an URL to the Schema Registry server.
+                // This is only required if you are importing Avro records. It is used to retrieve Avro schemas of Kafka
+                // topics. Avro is set as default record value format.
+                " TOPIC_NAME = '" + kafkaSetup.getTopicName() + "'\n" + //
+                " TABLE_NAME = '" + targetTable.getFullyQualifiedName() + "'\n" + //
+                " GROUP_ID = 'exasol-kafka-udf-consumers'\n";
+        LOGGER.info("Executing query '" + sql + "'");
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java:285]:
```suggestion
```
Already logged in `executeStatement()`

File: src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            final short replicationFactor = 1;
+            final NewTopic newTopic = new NewTopic(topicName, partitions, replicationFactor);
+
+            final CreateTopicsResult result = admin.createTopics(Collections.singleton(newTopic));
+
+            final KafkaFuture<Void> future = result.values().get(topicName);
+            final var createTopicsResult = future.get();
+        } catch (final Exception ex) {
+            LOG.warning("Exception occurred during Kafka topic creation: '" + ex.getMessage() + "'");
+        }
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java:59]: Re-throw exception

File: src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            producerProps.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
+
+            final Producer<String, String> producer = new KafkaProducer<>(producerProps);
+            // for (int i = 0; i < 100; i++)
+            producer.send(new ProducerRecord<String, String>(topicName, Integer.toString(1), "OK"));
+            producer.send(new ProducerRecord<String, String>(topicName, Integer.toString(2), "WARN"));
+            producer.close();
+        } catch (final Exception ex) {
+            LOG.warning("Exception occurred producing Kafka records: '" + ex.getMessage() + "'");
+        }
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java:79]: Re-throw exception

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: extension/src/test-utils.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+}
+export function adapterScript({ name = "S3_FILES_ADAPTER", type = "ADAPTER", text = "adapter script" }: Partial<ExaScriptsRow>): ExaScriptsRow {
+    return script({ name, type, text })
+}
+export function importScript({ name = "IMPORT_FROM_S3_DOCUMENT_FILES", type = "UDF", inputType = "SET", resultType = "EMITS" }: Partial<ExaScriptsRow>): ExaScriptsRow {
+    return script({ name, type, inputType, resultType })
+}
+export function scriptWithVersion(name: string, version: string): ExaScriptsRow {
+    return script({ name, text: `CREATE ... %jar /path/to/exasol-kafka-connector-extension-${version}.jar; more text` })
+}
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply extension/src/test-utils.ts:67]: Can be deleted, created a commit.

File: extension/src/extension.test.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        }
+        function script({ schema = "schema", name = "name", inputType, resultType = "EMITS", type = "UDF", text = "", comment }: Partial<ExaScriptsRow>): ExaScriptsRow {
+            return { schema, name, inputType, resultType, type, text, comment }
+        }
+        function setScript(name: string, className: string, version = EXTENSION_DESCRIPTION.version): ExaScriptsRow {
+            return script({ name, inputType: "SET", text: text(name, className, version) })
+        }
+        function scalarScript(name: string, className: string, version = EXTENSION_DESCRIPTION.version): ExaScriptsRow {
+            return script({ name, inputType: "SCALAR", text: text(name, className, version) })
+        }
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply extension/src/extension.test.ts]: Removed.

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+            ExecutionException, InterruptedException {
+        exasolTestSetup = new ExasolTestSetupFactory(Paths.get("no-cloud-setup")).getTestSetup();
+        setup = ExtensionManagerSetup.create(exasolTestSetup, ExtensionBuilder.createDefaultNpmBuilder(
+                EXTENSION_SOURCE_DIR, EXTENSION_SOURCE_DIR.resolve("dist").resolve(EXTENSION_ID)));
+        exasolTestSetup.getDefaultBucket().uploadFile(ADAPTER_JAR, ADAPTER_JAR.getFileName().toString());
+        client = setup.client();
+        kafkaSetup = KafkaTestSetup.create();
+        connection = exasolTestSetup.createConnection();
+        exasolObjectFactory = new ExasolObjectFactory(connection,
+                ExasolObjectConfiguration.builder().withJvmOptions("-Dcom.amazonaws.sdk.disableCbor=true").build());
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java:64]: After our discussion around debugging these, I agree, I've decided to remove it.

File: extension/src/extension.test.ts [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            const executeCalls = context.mocks.sqlExecute.mock.calls
+            expect(executeCalls.length).toBe(10)
+
+            const expectedScriptNames = ["KAFKA_METADATA", "KAFKA_IMPORT", "KAFKA_CONSUMER"]
+
+            const createScriptStatements = executeCalls.slice(0, 5).map(args => args[0])
+            const createCommentStatements = executeCalls.slice(5, 10).map(args => args[0])
+
+            expect(createScriptStatements).toHaveLength(5)
+            expect(createCommentStatements).toHaveLength(5)
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply extension/src/extension.test.ts]: Done. It didn't 😛

File: pom.xml [hunk 10]
```diff
@@ -3,23 +3,19 @@
...
-    <parent>
-        <artifactId>kafka-connector-extension-generated-parent</artifactId>
-        <groupId>com.exasol</groupId>
-        <version>1.6.0</version>
-        <relativePath>pk_generated_parent.pom</relativePath>
-    </parent>
     <properties>
         <!-- Update this once the Avro4s version is updated -->
         <!-- com.sksamuel.avro4s:avro4s-core_2.13:4.1.1 requires scala version: 2.13.3 -->
+        <java.version>11</java.version>
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply pom.xml:13]: Won't touch this as it breaks the build for me locally

File: pom.xml [hunk 11]
```diff
@@ -194,6 +196,23 @@
...
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>kafka</artifactId>
+            <version>1.19.0</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.kafka</groupId>
+            <artifactId>kafka-clients</artifactId>
+            <version>3.5.1</version>
```
[the remark this answers is not in the document]
Comment (pj-spoelders) [reply pom.xml:214]: if I do this I break the build since they're also used in the kafka-connector itself.

File: .project-keeper.yml [hunk 12]
```diff
@@ -4,3 +4,5 @@ sources:
     modules:
       - jar_artifact
       - integration_tests
+excludes:
+  - "E-PK-CORE-18: Outdated content: '.github/workflows/ci-build.yml'"
```
Comment (kaklakariada) [inline .project-keeper.yml:8]: Remove `.github/workflows/ci-build.yml` from `.gitattributes`

File: doc/changes/changes_1.7.0.md [hunk 13]
```diff
@@ -1,32 +1,58 @@
...
+    * org.eclipse.jetty.http2:http2-common:jar:9.4.51.v20230217:test
+        * CVE-2023-44487
+    * org.apache.avro:avro:jar:1.9.2:test 
+        * CVE-2023-39410
+    * org.eclipse.jetty.http2:http2-hpack:jar:9.4.51.v20230217:test
+        * CVE-2023-36478
+    * org.apache.zookeeper:zookeeper:jar:3.6.3:test
+        * CVE-2023-44981
+    * org.json:json:jar:20230227:test
+        * CVE-2023-5072
```
Comment (kaklakariada) [inline doc/changes/changes_1.7.0.md:33]: Are there versions available that fix these vulnerabilities?

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java [hunk 14]
```diff
@@ -246,20 +245,23 @@ private void verifyImportWorks() {
 
             final Table targetTable = schema.createTableBuilder("TARGET")
 
-                    .column("SENSOR_ID", "INTEGER").column("STATUS", "VARCHAR(10)") //
-                    .column("KAFKA_PARTITION", "DECIMAL(18, 0)").column("KAFKA_OFFSET", "DECIMAL(36, 0)").build();
+                    //.column("SENSOR_ID", "INTEGER")//
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java:248]:
```suggestion
```

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java [hunk 15]
```diff
@@ -246,20 +245,23 @@ private void verifyImportWorks() {
...
+                    .column("STATUS", "VARCHAR(10)") //
+                    .column("KAFKA_PARTITION", "DECIMAL(18, 0)")//
+                    .column("KAFKA_OFFSET", "DECIMAL(36, 0)")//
+                    .build();
             // CREATE CONNECTION (optional, see
             // https://github.com/exasol/kafka-connector-extension/blob/main/doc/user_guide/user_guide.md#importing-records)
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java:254]:
```suggestion
```

File: src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java [hunk 16]
```diff
@@ -53,37 +63,43 @@ static KafkaTestSetup create() throws ExecutionException, InterruptedException {
...
+            future.get();
+
+            //final Set<String> topicNames = admin.listTopics().names().get();
+
+            LOG.info("Succesfully created topic");
         } catch (final Exception ex) {
             LOG.warning("Exception occurred during Kafka topic creation: '" + ex.getMessage() + "'");
+            throw ex;
         }
+        // PRODUCE
```
Comment (kaklakariada) [inline src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java:75]: Extract the code to produce entries into a separate method and call it from the test when running the test.
Else only these two entries are produced and when additional tests (e.g. upgrade test) also read entries they won't find any.

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: doc/changes/changes_1.7.0.md (hunk 13 above)
Comment (pj-spoelders) [reply doc/changes/changes_1.7.0.md:33]: I've updated everything I could in the project itself. I'll have a closer look at our own depedencies down the line, see what we can update.

**Review by kaklakariada — APPROVED** [review]

File: .project-keeper.yml (hunk 12 above)
Comment (pj-spoelders) [reply .project-keeper.yml]: done

File: src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java (hunk 16 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/cloudetl/kafka/KafkaTestSetup.java]: done

File: src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java (hunk 15 above)
Comment (pj-spoelders) [reply src/test/java/com/exasol/cloudetl/kafka/ExtensionIT.java]: removed

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**